perception/multi_object_tracker/src/processor/input_manager.cpp

```diff
@@ -0,0 +1,342 @@
+
+  latest_measurement_time_ = node_.now();
+  latest_message_time_ = node_.now();
+}
+
+bool InputStream::getTimestamps(
+  rclcpp::Time & latest_measurement_time, rclcpp::Time & latest_message_time) const
+{
+  if (!isTimeInitialized()) {
+    return false;
+  }
+  latest_measurement_time = latest_measurement_time_;
+  latest_message_time = latest_message_time_;
+  return true;
+}
+
+void InputStream::onMessage(
+  const autoware_auto_perception_msgs::msg::DetectedObjects::ConstSharedPtr msg)
+{
+  const DetectedObjects objects = *msg;
+  objects_que_.push_back(objects);
+  if (objects_que_.size() > que_size_) {
+    objects_que_.pop_front();
+  }
+
+  // update the timing statistics
+  rclcpp::Time now = node_.now();
+  rclcpp::Time objects_time(objects.header.stamp);
+  updateTimingStatus(now, objects_time);
+
+  // trigger the function if it is set
+  if (func_trigger_) {
+    func_trigger_(index_);
+  }
+}
+
+void InputStream::updateTimingStatus(const rclcpp::Time & now, const rclcpp::Time & objects_time)
+{
+  // Update latency statistics
+  // skip initial messages for the latency statistics
+  if (initial_count_ > 4) {
+    const double latency = (now - objects_time).seconds();
+    if (initial_count_ < 16) {
+      // set higher gain for the initial messages
+      constexpr double initial_gain = 0.5;
+      latency_mean_ = (1.0 - initial_gain) * latency_mean_ + initial_gain * latency;
+    } else {
+      constexpr double gain = 0.05;
+      latency_mean_ = (1.0 - gain) * latency_mean_ + gain * latency;
+      const double latency_delta = latency - latency_mean_;
+      latency_var_ = (1.0 - gain) * latency_var_ + gain * latency_delta * latency_delta;
+    }
+  }
+
+  // Calculate interval, Update interval statistics
+  if (initial_count_ > 4) {
+    const double interval = (now - latest_message_time_).seconds();
+    if (interval < 0.0) {
+      RCLCPP_WARN(
+        node_.get_logger(),
+        "InputManager::updateTimingStatus %s: Negative interval detected, now: %f, "
+        "latest_message_time_: %f",
+        long_name_.c_str(), now.seconds(), latest_message_time_.seconds());
+    } else if (initial_count_ < 24) {
+      // Initialization
+      constexpr double initial_gain = 0.5;
+      interval_mean_ = (1.0 - initial_gain) * interval_mean_ + initial_gain * interval;
+    } else {
+      // The interval is considered regular if it is within 0.5 and 1.5 times the mean interval
+      bool update_statistics = interval > 0.5 * interval_mean_ && interval < 1.5 * interval_mean_;
+      if (update_statistics) {
+        constexpr double gain = 0.05;
+        interval_mean_ = (1.0 - gain) * interval_mean_ + gain * interval;
+        const double interval_delta = interval - interval_mean_;
+        interval_var_ = (1.0 - gain) * interval_var_ + gain * interval_delta * interval_delta;
+      }
+    }
+  }
+
+  // Update time
+  latest_message_time_ = now;
+  if (std::abs((latest_measurement_time_ - objects_time).seconds()) > 3.0) {
```

Comment:
As much as possible, I want to avoid hardcoding values and use variables instead.
```
// example
const float delay_threshold = 3.0; /* sec */
```

Reply:
The threshold is expressed as your suggestion. 
Thank you.

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/6820/commits/4445c45fb9bcd9159683a24afa0cf7d259aa45b2